### PR TITLE
Obs not voided when updated via saving an Encounter - TRUNK-50

### PR DESCRIPTION
**Pull request by wluyima**
Implemented Obs.osDirty and added ObsBehaviorTest - TRUNK-50

Fixed failing unit tests - TRUNK-50

Cleaned up code - TRUNK-50

Fixed more tests- TRUNK-50

Cleaned up code - TRUNK-50

# Conflicts:
#	api/src/test/java/org/openmrs/api/EncounterServiceTest.java

### CODE REVIEW

File: api/src/test/java/org/openmrs/ObsTest.java [hunk 1]
```diff
@@ -554,4 +646,272 @@ public void setFormField_shouldRejectAPathContainingTheSeparator() throws Except
...
+			assertFalse("Failed to rest Obs.dirty after testing field: " + fieldName, obs.isDirty());
+		}
+	}
+	
+	/**
+	 * @see Obs#isDirty()
+	 * @verifies return false when only mutable fields are changed
+	 */
+	@Test
+	public void isDirty_shouldReturnFalseWhenOnlyMutableFieldsAreChanged() throws Exception {
```
Comment (rkorytkowski) [inline api/src/test/java/org/openmrs/ObsTest.java]: Codacy Issue found: [Method names should not contain underscores](https://www.codacy.com/app/openmrs/openmrs-core/file/2318734355/issues/source?bid=3204099&fileBranchId=3204127#l711)

File: api/src/test/java/org/openmrs/ObsTest.java [hunk 2]
```diff
@@ -554,4 +646,272 @@ public void setFormField_shouldRejectAPathContainingTheSeparator() throws Except
...
+		obs.setComment("some non null value");
+		assertTrue(obs.isDirty());
+	}
+	
+	/**
+	 * @see Obs#setFormField(String,String)
+	 * @verifies not mark the obs as dirty when the value has not been changed
+	 */
+	@Test
+	public void setFormField_shouldNotMarkTheObsAsDirtyWhenTheValueHasNotBeenChanged() throws Exception {
```
Comment (rkorytkowski) [inline api/src/test/java/org/openmrs/ObsTest.java]: Codacy Issue found: [Method names should not contain underscores](https://www.codacy.com/app/openmrs/openmrs-core/file/2318734355/issues/source?bid=3204099&fileBranchId=3204127#l749)

File: api/src/test/java/org/openmrs/ObsTest.java [hunk 3]
```diff
@@ -554,4 +646,272 @@ public void setFormField_shouldRejectAPathContainingTheSeparator() throws Except
...
+		BeanUtils.copyProperties(obs, BeanUtils.cloneBean(obs));
+		assertFalse(obs.isDirty());
+	}
+	
+	/**
+	 * @see Obs#isDirty()
+	 * @verifies return true when any immutable field has been changed
+	 */
+	@Test
+	public void isDirty_shouldReturnTrueWhenAnyImmutableFieldHasBeenChanged() throws Exception {
```
Comment (rkorytkowski) [inline api/src/test/java/org/openmrs/ObsTest.java]: Codacy Issue found: [Method names should not contain underscores](https://www.codacy.com/app/openmrs/openmrs-core/file/2318734355/issues/source?bid=3204099&fileBranchId=3204127#l672)

File: api/src/test/java/org/openmrs/ObsTest.java [hunk 4]
```diff
@@ -554,4 +646,272 @@ public void setFormField_shouldRejectAPathContainingTheSeparator() throws Except
...
+		obs.setFormField(obs.getFormFieldNamespace(), obs.getFormFieldPath());
+		assertFalse(obs.isDirty());
+	}
+	
+	/**
+	 * @see Obs#setFormField(String,String)
+	 * @verifies mark the obs as dirty when the value has been changed
+	 */
+	@Test
+	public void setFormField_shouldMarkTheObsAsDirtyWhenTheValueHasBeenChanged() throws Exception {
```
Comment (rkorytkowski) [inline api/src/test/java/org/openmrs/ObsTest.java]: Codacy Issue found: [Method names should not contain underscores](https://www.codacy.com/app/openmrs/openmrs-core/file/2318734355/issues/source?bid=3204099&fileBranchId=3204127#l760)

File: api/src/test/java/org/openmrs/ObsTest.java [hunk 5]
```diff
@@ -554,4 +646,272 @@ public void setFormField_shouldRejectAPathContainingTheSeparator() throws Except
...
+		obs.setComment(null);
+		assertTrue(obs.isDirty());
+	}
+	
+	/**
+	 * @see Obs#isDirty()
+	 * @verifies return true when a field is changed from a null to a non null value
+	 */
+	@Test
+	public void isDirty_shouldReturnTrueWhenAnImmutableFieldIsChangedFromANullToANonNullValue() throws Exception {
```
Comment (rkorytkowski) [inline api/src/test/java/org/openmrs/ObsTest.java]: Codacy Issue found: [Method names should not contain underscores](https://www.codacy.com/app/openmrs/openmrs-core/file/2318734355/issues/source?bid=3204099&fileBranchId=3204127#l737)

File: api/src/test/java/org/openmrs/ObsTest.java [hunk 6]
```diff
@@ -554,4 +646,272 @@ public void setFormField_shouldRejectAPathContainingTheSeparator() throws Except
...
+		obs.setDateVoided(new Date());
+		assertFalse(obs.isDirty());
+	}
+	
+	/**
+	 * @see Obs#isDirty()
+	 * @verifies return true when a field is changed from a non null to a null value
+	 */
+	@Test
+	public void isDirty_shouldReturnTrueWhenAnImmutableFieldIsChangedFromANonNullToANullValue() throws Exception {
```
Comment (rkorytkowski) [inline api/src/test/java/org/openmrs/ObsTest.java]: Codacy Issue found: [Method names should not contain underscores](https://www.codacy.com/app/openmrs/openmrs-core/file/2318734355/issues/source?bid=3204099&fileBranchId=3204127#l725)

File: api/src/test/java/org/openmrs/ObsTest.java [hunk 7]
```diff
@@ -554,4 +646,272 @@ public void setFormField_shouldRejectAPathContainingTheSeparator() throws Except
 		Obs obs = new Obs();
 		obs.setFormField("", path);
 	}
+	
+	/**
+	 * @see Obs#isDirty()
+	 * @verifies return false when no change has been made
+	 */
+	@Test
+	public void isDirty_shouldReturnFalseWhenNoChangeHasBeenMade() throws Exception {
```
Comment (rkorytkowski) [inline api/src/test/java/org/openmrs/ObsTest.java]: Codacy Issue found: [Method names should not contain underscores](https://www.codacy.com/app/openmrs/openmrs-core/file/2318734355/issues/source?bid=3204099&fileBranchId=3204127#l655)

File: api/src/main/java/org/openmrs/Obs.java [hunk 8]
```diff
@@ -435,6 +458,8 @@ public void setGroupMembers(Set<Obs> groupMembers) {
 	 * @param member Obs to add to this group
 	 * @see #setGroupMembers(Set)
 	 * @see #getGroupMembers()
+	 * @should return true when a new obs is added as a member
```
Comment (dkayiwa) [inline api/src/main/java/org/openmrs/Obs.java]: This method does no return a boolean. It is instead void.

File: api/src/main/java/org/openmrs/Obs.java [hunk 9]
```diff
@@ -462,6 +489,8 @@ public void addGroupMember(Obs member) {
 	 * @param member Obs to remove from this group
 	 * @see #setGroupMembers(Set)
 	 * @see #getGroupMembers()
+	 * @should return true when an obs is removed
```
Comment (dkayiwa) [inline api/src/main/java/org/openmrs/Obs.java]: This method does no return a boolean. It is instead void.

File: api/src/main/java/org/openmrs/Obs.java [hunk 10]
```diff
@@ -1190,6 +1252,31 @@ public void setFormField(String namespace, String formFieldPath) {
...
+	 * @should return false when only mutable fields are changed
+	 * @should return true when an immutable field is changed from a null to a non null value
+	 * @should return true when an immutable field is changed from a non null to a null value
+	 */
+	public boolean isDirty() {
+		return dirty;
+	}
+	
+	private void markAsDirty(Object oldValue, Object newValue) {
+		//Should we ignore the case for Strings?
```
Comment (dkayiwa) [inline api/src/main/java/org/openmrs/Obs.java]: What do you mean by the case for strings?

File: api/src/test/java/org/openmrs/ObsTest.java (hunk 1 above)
Comment (rkorytkowski) [inline api/src/test/java/org/openmrs/ObsTest.java]: Please ignore this kind of issues. It is a misconfigured analysis rule.

File: api/src/main/java/org/openmrs/api/handler/OpenmrsObjectSaveHandler.java [hunk 11]
```diff
@@ -82,7 +83,8 @@ public void handle(OpenmrsObject openmrsObject, User creator, Date dateCreated,
 			}
 			
 			//We are dealing with only strings
-			if (!property.getPropertyType().equals(String.class)) {
+            //TODO We shouldn't be doing this for all immutable types and fields
```
Comment (dkayiwa) [inline api/src/main/java/org/openmrs/api/handler/OpenmrsObjectSaveHandler.java]: Do you plan to address this TODO?

File: api/src/main/java/org/openmrs/Obs.java (hunk 10 above)
Comment (wluyima) [inline api/src/main/java/org/openmrs/Obs.java]: As changing from lowercase to uppercase or vice versa

File: api/src/main/java/org/openmrs/api/handler/OpenmrsObjectSaveHandler.java (hunk 11 above)
Comment (wluyima) [inline api/src/main/java/org/openmrs/api/handler/OpenmrsObjectSaveHandler.java]: Nope, we should probably have a ticket for it

File: api/src/main/java/org/openmrs/Obs.java (hunk 10 above)
Comment (dkayiwa) [inline api/src/main/java/org/openmrs/Obs.java]: @wluyima do you know why bamboo is failing with this commit?
It passes locally and on travis.

File: api/src/main/java/org/openmrs/Obs.java (hunk 8 above)
Comment (wluyima) [inline api/src/main/java/org/openmrs/Obs.java]: Correct, you can rephrase it, it should be something like '@should mark the obs as dirty when a new member is added'

**Pull request merged**